pkg/controllers/corecluster/corecluster_controller.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package corecluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+	"github.com/openshift/cluster-capi-operator/pkg/util"
+)
+
+const (
+	controllerName                    = "CoreClusterController"
+	capiInfraClusterAPIVersionV1Beta1 = "infrastructure.cluster.x-k8s.io/v1beta1"
+	capiInfraClusterAPIVersionV1Beta2 = "infrastructure.cluster.x-k8s.io/v1beta2"
+	clusterOperatorName               = "cluster-api"
+)
+
+var (
+	errPlatformStatusShouldNotBeNil                         = errors.New("infrastructure PlatformStatus should not be nil")
+	errMappingOpenshiftPlatformToInfraClusterKindAndVersion = errors.New("failed to map Infrastructure PlatformType to a Cluster API infra cluster kind, platform not supported")
+	errOpenshiftInfraShouldNotBeNil                         = errors.New("infrastructure object should not be nil")
+	errOpenshiftInfrastructureNameShouldNotBeEmpty          = errors.New("infrastructure object's status.infrastructureName should not be empty")
```

Comment:
There's inconsistencies in these messages, perhaps we should
```suggestion
	errPlatformStatusShouldNotBeNil                         = errors.New("infrastructure platformStatus should not be nil")
	errMappingOpenshiftPlatformToInfraClusterKindAndVersion = errors.New("failed to map infrastructure platformType to infrastructure cluster kind, platform not supported")
	errOpenshiftInfraShouldNotBeNil                         = errors.New("infrastructure object should not be nil")
	errOpenshiftInfrastructureNameShouldNotBeEmpty          = errors.New("infrastructure object's infrastructureName should not be empty")
```